pennylane/plugins/expt_tensornet_tf.py

```diff
@@ -0,0 +1,360 @@
+                )
+            stat_node = np.zeros(tuple([2] * len(wires)))
+            stat_node[tuple(par[0])] = 1
+            self._state_node.tensor = tf.constant(stat_node, dtype=tf.complex128)
+            return
+
+        A = self._get_operator_matrix(operation, par)
+        num_mult_idxs = len(wires)
+        A = tf.reshape(A, [2] * num_mult_idxs * 2)
+        op_node = self._add_node(A, wires=wires, name=operation)
+        for idx, w in enumerate(wires):
+            self._add_edge(op_node, num_mult_idxs + idx, self._state_node, w)
+            self._free_edges[w] = op_node[idx]
+        # TODO: can be smarter here about collecting contractions?
+        self._state_node = tn.contract_between(
+            op_node, self._state_node, output_edge_order=self._free_edges
+        )
+
+    def ev(self, obs_nodes, wires):
+        r"""Expectation value of observables on specified wires.
+
+         Args:
+            obs_nodes (Sequence[tn.Node]): the observables as tensornetwork Nodes
+            wires (Sequence[Sequence[int]]): measured subsystems for each observable
+         Returns:
+            float: expectation value :math:`\expect{A} = \bra{\psi}A\ket{\psi}`
+        """
+
+        all_wires = tuple(w for w in range(self.num_wires))
+        ket = self._add_node(self._state_node, wires=all_wires, name="Ket")
+        bra = self._add_node(tn.conj(ket), wires=all_wires, name="Bra")
+        meas_wires = []
+        # We need to build up <psi|A|psi> step-by-step.
+        # For wires which are measured, we need to connect edges between
+        # bra, obs_node, and ket.
+        # For wires which are not measured, we need to connect edges between
+        # bra and ket.
+        # We use the convention that the indices of a tensor are ordered like
+        # [output_idx1, output_idx2, ..., input_idx1, input_idx2, ...]
+        for obs_node, obs_wires in zip(obs_nodes, wires):
+            meas_wires.extend(obs_wires)
+            for idx, w in enumerate(obs_wires):
+                output_idx = idx
+                input_idx = len(obs_wires) + idx
+                self._add_edge(obs_node, input_idx, ket, w)  # A|psi>
+                self._add_edge(bra, w, obs_node, output_idx)  # <psi|A
+        for w in set(all_wires) - set(meas_wires):
+            self._add_edge(bra, w, ket, w)  # |psi[w]|**2
+
+        # At this stage, all nodes are connected, and the contraction yields a
+        # scalar value.
+        contracted_ket = ket
+        for obs_node in obs_nodes:
+            contracted_ket = tn.contract_between(obs_node, contracted_ket)
+        expval = tn.contract_between(bra, contracted_ket).tensor
+
+        # TODO: add complex non-vanishing part warning
+
+        return tf.math.real(expval)
+
+    def execute(self, queue, observables, parameters=None):
+        self.check_validity(queue, observables)
+
+        results = []
+
+        self.tape = tf.GradientTape(persistent=True)
+
+        with self.tape:
```

Comment:
As far as I can tell, we have to overwrite `execute()`, as we need to:

* Convert the final TF tensor into a NumPy array so that the QNode gets a type it recognizes
* Capture the end tensor before it is converted to a NumPy array.

If we modify the device API, it may be possible to avoid this. But to avoid premature optimization, and since this is experimental, I think overwriting `execute()` is better for now(?). 

There is an ADR currently that addresses shortfalls in the device API, we should revisit the structure of this plugin once that ADR is decided, and when this plugin is promoted to core.

Reply:
Done, now using the pre-existing context